### PR TITLE
ICU-20739 Force seconds if the skeleton has fractional seconds

### PR DESCRIPTION
**Pull request by mihnita**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20739
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java [hunk 1]
```diff
@@ -2649,6 +2649,25 @@ DateTimeMatcher set(String pattern, FormatParser fp, boolean allowDuplicateField
...
+                // Force the use of seconds
+                for (int i = 0; i < types.length; ++i) {
+                    int[] row = types[i];
+                    if (row[1] == SECOND) {
+                        // first entry for SECOND
+                        original.populate(SECOND, (char)row[0], (char)row[3]);
+                        baseOriginal.populate(SECOND, (char)row[0], (char)row[3]);
+                        // We add value.length, same as above, when type is first initialized.
+                        // The value we want to "fake" here is "s", and 1 means "s".length()
+                        int subField = row[2];
```
Comment (sffc) [inline icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java]: According to the `types` table at the bottom of the file, `row[2]` is `NUMERIC`, which is defined as `0x100`, so I think you can just assign `type[SECOND] = row[2]` directly, like they're doing in the next loop with `DAYPERIOD`.  Also makes it more consistent with C++.

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java [hunk 2]
```diff
@@ -2649,6 +2649,25 @@ DateTimeMatcher set(String pattern, FormatParser fp, boolean allowDuplicateField
...
             }
+
+            // #20739, we have a skeleton with milliseconde, but no seconds
+            if (!original.isFieldEmpty(FRACTIONAL_SECOND) && original.isFieldEmpty(SECOND)) {
+                // Force the use of seconds
+                for (int i = 0; i < types.length; ++i) {
+                    int[] row = types[i];
+                    if (row[1] == SECOND) {
+                        // first entry for SECOND
+                        original.populate(SECOND, (char)row[0], (char)row[3]);
```
Comment (sffc) [inline icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java]: I wouldn't nitpick about this if I weren't already leaving another comment, but you only need to cast `row[0]`, not `row[3]`.

Comment (mihnita) [reply icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java]: I'll fix it if you think I should fix the first one.

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java (hunk 1 above)
Comment (mihnita) [reply icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java]: Made C++ consistent with Java (and itself)
Without adding that length the "mS" Java the unit tests fails.

**Comment by sffc** [comment]
LGTM please squash

**Pull request merged**